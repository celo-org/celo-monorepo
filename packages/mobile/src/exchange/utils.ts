import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
