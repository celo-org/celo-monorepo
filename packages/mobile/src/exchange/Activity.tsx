import * as React from 'react'
import { connect } from 'react-redux'
import UserTransactionsQuery from 'src/apollo/types'
import { transactionQuery } from 'src/home/TransactionsList'
import { RootState } from 'src/redux/reducers'
import { resetStandbyTransactions } from 'src/transactions/actions'
import { StandbyTransaction, TransactionTypes } from 'src/transactions/reducer'
import TransactionFeed, { FeedType } from 'src/transactions/TransactionFeed'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

interface DispatchProps {
  resetStandbyTransactions: typeof resetStandbyTransactions
}

interface StateProps {
  address?: string | null
}

type Props = StateProps & DispatchProps

const mapStateToProps = (state: RootState): StateProps => ({
  address: currentAccountSelector(state),
})

function filterToExchangeTxs(tx: StandbyTransaction) {
  return tx.type === TransactionTypes.EXCHANGE
}

export class Activity extends React.Component<Props> {
  componentDidMount() {
    this.props.resetStandbyTransactions()
    Logger.info('Activity feed', JSON.stringify(transactionQuery))
  }

  render() {
    const { address } = this.props
    const queryAddress = address || ''
    return (
      <UserTransactionsQuery
        query={transactionQuery}
        pollInterval={1000}
        variables={{ address: queryAddress }}
      >
        {({ loading, error, data }) => {
          return (
            <TransactionFeed
              kind={FeedType.EXCHANGE}
              loading={loading}
              error={error}
              data={data}
              transactionFilter={filterToExchangeTxs}
            />
          )
        }}
      </UserTransactionsQuery>
    )
  }
}

export default connect<StateProps, DispatchProps, {}, RootState>(
  mapStateToProps,
  {
    resetStandbyTransactions,
  }
)(Activity)
