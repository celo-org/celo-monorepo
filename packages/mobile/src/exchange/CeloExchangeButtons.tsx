// VIEW which contains buy and sell buttons for CELO.

import Button, { BtnSizes, BtnTypes } from '@celo/react-components/components/Button'
import { CURRENCY_ENUM } from '@celo/utils'
import { StackNavigationProp } from '@react-navigation/stack'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, View } from 'react-native'
import { useSelector } from 'react-redux'
import { CeloExchangeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { DOLLAR_TRANSACTION_MIN_AMOUNT, GOLD_TRANSACTION_MIN_AMOUNT } from 'src/config'
import { exchangeRatePairSelector } from 'src/exchange/reducer'
import { celoTokenBalanceSelector } from 'src/goldToken/selectors'
import { Namespaces } from 'src/i18n'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import { stableTokenBalanceSelector } from 'src/stableToken/reducer'

interface Props {
  navigation: StackNavigationProp<StackParamList, Screens.ExchangeHomeScreen>
}

export default function CeloExchangeButtons({ navigation }: Props) {
  const { t } = useTranslation(Namespaces.exchangeFlow9)

  const dollarBalance = useSelector(stableTokenBalanceSelector)
  const goldBalance = useSelector(celoTokenBalanceSelector)
  const exchangeRate = useSelector(exchangeRatePairSelector)

  const hasDollars = new BigNumber(dollarBalance || 0).isGreaterThan(DOLLAR_TRANSACTION_MIN_AMOUNT)
  const hasGold = new BigNumber(goldBalance || 0).isGreaterThan(GOLD_TRANSACTION_MIN_AMOUNT)

  function goToBuyGold() {
    ValoraAnalytics.track(CeloExchangeEvents.celo_home_buy)
    navigation.navigate(Screens.ExchangeTradeScreen, {
      makerTokenDisplay: {
        makerToken: CURRENCY_ENUM.DOLLAR,
        makerTokenBalance: dollarBalance || '0',
      },
    })
  }

  function goToBuyDollars() {
    ValoraAnalytics.track(CeloExchangeEvents.celo_home_sell)
    navigation.navigate(Screens.ExchangeTradeScreen, {
      makerTokenDisplay: {
        makerToken: CURRENCY_ENUM.GOLD,
        makerTokenBalance: goldBalance || '0',
      },
    })
  }

  if (!exchangeRate || (!hasDollars && !hasGold)) {
    return <View style={styles.emptyContainer} />
  }

  return (
    <View style={styles.buttonContainer}>
      {hasDollars && (
        <Button
          text={t('buy')}
          size={BtnSizes.FULL}
          onPress={goToBuyGold}
          style={styles.button}
          type={BtnTypes.TERTIARY}
          testID="BuyCelo"
        />
      )}
      {hasGold && (
        <Button
          size={BtnSizes.FULL}
          text={t('sell')}
          onPress={goToBuyDollars}
          style={styles.button}
          type={BtnTypes.TERTIARY}
          testID="SellCelo"
        />
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  emptyContainer: {
    marginTop: 24,
  },
  buttonContainer: {
    flexDirection: 'row',
    flex: 1,
    marginTop: 24,
    marginBottom: 28,
    marginHorizontal: 12,
  },
  button: {
    marginHorizontal: 4,
    flex: 1,
  },
})
