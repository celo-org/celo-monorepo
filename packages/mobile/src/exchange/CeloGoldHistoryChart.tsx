import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import BigNumber from 'bignumber.js'
import _ from 'lodash'
import React, { useCallback, useState } from 'react'
import { WithTranslation } from 'react-i18next'
import { ActivityIndicator, LayoutChangeEvent, StyleSheet, Text, View } from 'react-native'
import { Circle, G, Line, Text as SvgText } from 'react-native-svg'
import { useExchangeRate } from 'src/exchange/hooks'
import { exchangeHistorySelector } from 'src/exchange/reducer'
import { Namespaces, withTranslation } from 'src/i18n'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { convertDollarsToLocalAmount } from 'src/localCurrency/convert'
import { useLocalCurrencyCode } from 'src/localCurrency/hooks'
import { getLocalCurrencyExchangeRate } from 'src/localCurrency/selectors'
import useSelector from 'src/redux/useSelector'
import { goldToDollarAmount } from 'src/utils/currencyExchange'
import { getLocalCurrencyDisplayValue } from 'src/utils/formatting'
// @ts-ignore
import { VictoryGroup, VictoryLine, VictoryScatter } from 'victory-native'

const CHART_POINTS_NUMBER = 60
const CHART_WIDTH = variables.width - variables.contentPadding * 2
const CHART_HEIGHT = 180
const CHART_MIN_VERTICAL_RANGE = 0.1
const CHART_DOMAIN_PADDING = { y: [30, 30], x: [5, 5] }

interface OwnProps {
  testID?: string
}

type Props = WithTranslation & OwnProps

// ChartAwareSvgText draws text on the chart with avareness of its edges.
// Example: we want to draw some text at {x:10,y:10}(coordinates of the center).
// The text width is 33px and if draw it right away it will be cuted by the chart edges.
// The component will adjust coordinates to {x: (textWidthInPixels/2), y: 10}
function ChartAwareSvgText({
  position,
  x,
  y,
  value,
  chartWidth,
}: {
  position: 'top' | 'bottom'
  x: number
  y: number
  value: string
  chartWidth: number
}) {
  if (position === 'top') {
    y = y - 16
  } else if (position === 'bottom') {
    y = y + 25
  }
  const [adjustedX, setAdjustedX] = useState(x)
  const onLayout = useCallback(
    ({
      nativeEvent: {
        layout: { width },
      },
    }: LayoutChangeEvent) => {
      if (Math.abs(width - chartWidth) > 2) {
        if (x - width / 2 < 0) {
          setAdjustedX(width / 2)
        }
        if (x + width / 2 > chartWidth) {
          setAdjustedX(chartWidth - width / 2)
        }
      }
    },
    [x]
  )
  return (
    <SvgText onLayout={onLayout} fill="black" fontSize="14" x={adjustedX} y={y} textAnchor="middle">
      {value}
    </SvgText>
  )
}

function renderPointOnChart(
  chartData: Array<{ amount: number | BigNumber; displayValue: string }>
) {
  let lowestRateIdx = 0,
    highestRateIdx = 0
  chartData.forEach((rate, idx) => {
    if (rate.amount > chartData[highestRateIdx].amount) {
      highestRateIdx = idx
    }
    if (rate.amount < chartData[lowestRateIdx].amount) {
      lowestRateIdx = idx
    }
  })
  return ({ datum, x, y }: { x: number; y: number; datum: { _x: number; _y: number } }) => {
    const idx = datum._x
    switch (idx) {
      case chartData.length - 1:
        return (
          <G key={idx}>
            <Circle cx={x} cy={y} r="5" fill={'black'} />
            <Line x1={0} y1={y} x2={x} y2={y} stroke={colors.listBorder} strokeWidth="1" />
          </G>
        )

      case highestRateIdx:
        return (
          <ChartAwareSvgText
            x={x}
            y={y}
            key={idx}
            value={chartData[highestRateIdx].displayValue}
            position={'top'}
            chartWidth={CHART_WIDTH}
          />
        )

      case lowestRateIdx:
        return (
          <ChartAwareSvgText
            x={x}
            y={y}
            key={idx}
            value={chartData[lowestRateIdx].displayValue}
            position={'bottom'}
            chartWidth={CHART_WIDTH}
          />
        )

      default:
        return null
    }
  }
}

function Loader() {
  return (
    <View style={styles.loader}>
      <ActivityIndicator size="large" color={colors.celoGreen} />
    </View>
  )
}

function CeloGoldHistoryChart({ t, testID }: Props) {
  const calculateGroup = useCallback((er) => {
    return Math.floor(er.timestamp / (range / CHART_POINTS_NUMBER))
  }, [])

  const localCurrencyCode = useLocalCurrencyCode()
  const displayLocalCurrency = useCallback(
    (amount: BigNumber.Value) =>
      getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true),
    [localCurrencyCode]
  )
  const exchangeRate = useExchangeRate()
  const goldToDollars = useCallback((amount) => goldToDollarAmount(amount, exchangeRate), [
    exchangeRate,
  ])
  const localExchangeRate = useSelector(getLocalCurrencyExchangeRate)
  const dollarsToLocal = useCallback(
    (amount) => convertDollarsToLocalAmount(amount, localCurrencyCode ? localExchangeRate : 1),
    [localExchangeRate]
  )
  const [range] = useState(30 * 24 * 60 * 60 * 1000) // 30 days
  const exchangeHistory = useSelector(exchangeHistorySelector)

  if (!exchangeHistory.celoGoldExchangeRates.length) {
    return <Loader />
  }

  const groupedExchangeHistory = _.groupBy(exchangeHistory.celoGoldExchangeRates, calculateGroup)
  const latestGroup = calculateGroup(
    exchangeHistory.celoGoldExchangeRates[exchangeHistory.celoGoldExchangeRates.length - 1]
  )
  const chartData = _.range(
    Math.min(CHART_POINTS_NUMBER - 1, Object.keys(groupedExchangeHistory).length),
    0,
    -1
  ).map((i) => {
    const group = groupedExchangeHistory[latestGroup - i + 1]
    const localAmount = dollarsToLocal(
      group ? _.meanBy(group, (er) => parseFloat(er.exchangeRate)) : 0
    )
    return {
      amount: localAmount ? localAmount.toNumber() : 0,
      displayValue: localAmount ? displayLocalCurrency(localAmount) : '',
    }
  })

  const currentGoldRateInLocalCurrency = dollarsToLocal(goldToDollars(1))
  const oldestGoldRateInLocalCurrency = chartData[0].amount
  if (oldestGoldRateInLocalCurrency == null || currentGoldRateInLocalCurrency == null) {
    return <Loader />
  }
  // We need displayValue to show min/max on the chart. In case the
  // current value is min/max we do not need to show it once again,
  // therefor displayValue = ''
  chartData.push({ amount: currentGoldRateInLocalCurrency.toNumber(), displayValue: '' })
  const rateChange = currentGoldRateInLocalCurrency.minus(oldestGoldRateInLocalCurrency)
  const rateChangeInPercentage = currentGoldRateInLocalCurrency
    .div(oldestGoldRateInLocalCurrency)
    .minus(1)
    .multipliedBy(100)
  const RenderPoint = renderPointOnChart(chartData)

  const values = chartData.map((el) => el.amount)
  const min = Math.min(...values)
  const max = Math.max(...values)
  let domain = null
  // ensure that vertical chart range is at least CHART_MIN_VERTICAL_RANGE
  if (max - min < CHART_MIN_VERTICAL_RANGE) {
    const offset = CHART_MIN_VERTICAL_RANGE - (max - min) / 2
    domain = { y: [min - offset, max + offset] }
  }

  return (
    <View style={styles.container} testID={testID}>
      <View style={styles.goldPrice}>
        <View>
          <Text style={styles.goldPriceTitle}>{t('goldPrice')}</Text>
        </View>
        <View style={styles.goldPriceValues}>
          <Text style={styles.goldPriceCurrentValue}>
            {displayLocalCurrency(currentGoldRateInLocalCurrency)}
          </Text>
          <Text style={styles.goldPriceChange}>
            {rateChange.gt(0) ? '▴' : '▾'} {rateChange.toFixed(2)} (
            {rateChangeInPercentage.toFixed(2)}%)
          </Text>
        </View>
      </View>
      <VictoryGroup
        domainPadding={CHART_DOMAIN_PADDING}
        singleQuadrantDomainPadding={false}
        padding={{ left: 0 }}
        width={CHART_WIDTH}
        height={CHART_HEIGHT}
        data={chartData.map((el) => el.amount)}
        domain={domain}
      >
        <VictoryLine interpolation="monotoneX" />
        {/* 
        // @ts-ignore */}
        <VictoryScatter dataComponent={<RenderPoint />} />
      </VictoryGroup>
      <Text style={styles.timeframe}>{t('global:timeframes.30d')}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: variables.contentPadding,
  },
  goldPrice: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingTop: variables.contentPadding,
  },
  goldPriceTitle: {
    ...fontStyles.body,
    fontSize: 20,
    lineHeight: 28,
  },
  goldPriceValues: { alignItems: 'flex-end' },
  goldPriceCurrentValue: {
    fontSize: 24,
  },
  goldPriceChange: {
    ...fontStyles.body,
    ...fontStyles.semiBold,
    color: colors.celoGreen,
  },
  loader: {
    width: CHART_WIDTH + 32,
    height: CHART_HEIGHT + 130,
    alignItems: 'center',
    justifyContent: 'center',
  },
  timeframe: {
    textAlign: 'center',
    fontSize: 16,
    paddingBottom: 4,
    color: colors.gray,
  },
  chartStyle: {
    paddingRight: 0,
    paddingTop: 32,
    marginTop: -32,
    paddingBottom: 16,
  },
})

export default withTranslation(Namespaces.exchangeFlow9)(CeloGoldHistoryChart)
