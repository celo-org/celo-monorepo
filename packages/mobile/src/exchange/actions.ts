import BigNumber from 'bignumber.js'
import { ExchangeRatePair } from 'src/exchange/reducer'
import { CURRENCY_ENUM } from 'src/geth/consts'

export enum Actions {
  FETCH_EXCHANGE_RATE = 'EXCHANGE/FETCH_EXCHANGE_RATE',
  SET_EXCHANGE_RATE = 'EXCHANGE/SET_EXCHANGE_RATE',
  EXCHANGE_TOKENS = 'EXCHANGE/EXCHANGE_TOKENS',
}

export interface FetchExchangeRateAction {
  type: Actions.FETCH_EXCHANGE_RATE
  makerToken?: CURRENCY_ENUM
  makerAmount?: BigNumber
}

export const fetchExchangeRate = (
  makerToken?: CURRENCY_ENUM,
  makerAmount?: BigNumber
): FetchExchangeRateAction => ({
  type: Actions.FETCH_EXCHANGE_RATE,
  makerToken,
  makerAmount,
})

export interface SetExchangeRateAction {
  type: Actions.SET_EXCHANGE_RATE
  exchangeRatePair: ExchangeRatePair
}

export const setExchangeRate = (exchangeRatePair: ExchangeRatePair): SetExchangeRateAction => ({
  type: Actions.SET_EXCHANGE_RATE,
  exchangeRatePair,
})

export interface ExchangeTokensAction {
  type: Actions.EXCHANGE_TOKENS
  makerToken: CURRENCY_ENUM
  makerAmount: BigNumber
}

export const exchangeTokens = (
  makerToken: CURRENCY_ENUM,
  makerAmount: BigNumber
): ExchangeTokensAction => ({
  type: Actions.EXCHANGE_TOKENS,
  makerToken,
  makerAmount,
})
export type ActionTypes = SetExchangeRateAction | ExchangeTokensAction
