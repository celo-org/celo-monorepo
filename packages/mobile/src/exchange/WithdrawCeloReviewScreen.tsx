// SCREEN where users can review the data about the witdrawal of CELO they are about to make
// and confirm.

import Button, { BtnSizes, BtnTypes } from '@celo/react-components/components/Button'
import HorizontalLine from '@celo/react-components/components/HorizontalLine'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import { StackScreenProps } from '@react-navigation/stack'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch } from 'react-redux'
import { CeloExchangeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import CancelButton from 'src/components/CancelButton'
import LineItemRow from 'src/components/LineItemRow'
import ShortenedAddress from 'src/components/ShortenedAddress'
import { withdrawCelo } from 'src/exchange/actions'
import WithdrawCeloSummary from 'src/exchange/WithdrawCeloSummary'
import { CURRENCY_ENUM } from 'src/geth/consts'
import i18n, { Namespaces } from 'src/i18n'
import { emptyHeader, HeaderTitleWithBalance } from 'src/navigator/Headers'
import { navigate, navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { TopBarTextButton } from 'src/navigator/TopBarButton'
import { StackParamList } from 'src/navigator/types'
import useTypedSelector from 'src/redux/useSelector'
import DisconnectBanner from 'src/shared/DisconnectBanner'

type Props = StackScreenProps<StackParamList, Screens.WithdrawCeloReviewScreen>

function WithdrawCeloReviewScreen({ route }: Props) {
  const { amount, recipientAddress, feeEstimate } = route.params
  const { t } = useTranslation(Namespaces.exchangeFlow9)
  const isLoading = useTypedSelector((state) => state.exchange.isLoading)
  const dispatch = useDispatch()

  const onConfirmWithdraw = () => {
    ValoraAnalytics.track(CeloExchangeEvents.celo_withdraw_confirm, {
      amount: amount.toString(),
    })
    dispatch(withdrawCelo(amount, recipientAddress))
  }

  return (
    <SafeAreaView style={styles.container} edges={['bottom']}>
      <DisconnectBanner />
      <View style={styles.contentContainer}>
        <LineItemRow
          title={t('exchangeFlow9:withdrawCeloTo')}
          textStyle={fontStyles.regular}
          amount={<ShortenedAddress style={styles.withdrawAddress} address={recipientAddress} />}
        />
        <HorizontalLine />
        <WithdrawCeloSummary
          amount={amount}
          recipientAddress={recipientAddress}
          feeEstimate={feeEstimate}
        />
      </View>
      <Button
        onPress={onConfirmWithdraw}
        text={t(`global:withdraw`)}
        type={BtnTypes.TERTIARY}
        size={BtnSizes.FULL}
        style={styles.reviewBtn}
        testID="ConfirmWithdrawButton"
        showLoading={isLoading}
        loadingColor={colors.light}
      />
    </SafeAreaView>
  )
}

WithdrawCeloReviewScreen.navigationOptions = () => {
  const onCancel = () => {
    ValoraAnalytics.track(CeloExchangeEvents.celo_withdraw_cancel)
    navigate(Screens.ExchangeHomeScreen)
  }
  const onEdit = () => {
    ValoraAnalytics.track(CeloExchangeEvents.celo_withdraw_edit)
    navigateBack()
  }
  return {
    ...emptyHeader,
    headerTitle: () => (
      <HeaderTitleWithBalance
        title={i18n.t('exchangeFlow9:withdrawCeloReview')}
        token={CURRENCY_ENUM.GOLD}
      />
    ),
    headerLeft: () => <CancelButton onCancel={onCancel} />,
    headerRight: () => (
      <TopBarTextButton
        title={i18n.t('global:edit')}
        testID="EditButton"
        onPress={onEdit}
        titleStyle={{ color: colors.goldDark }}
        eventName={CeloExchangeEvents.celo_sell_edit}
      />
    ),
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  contentContainer: {
    flex: 1,
    marginTop: 38,
    paddingHorizontal: 16,
  },
  withdrawAddress: {
    ...fontStyles.regular,
    fontSize: 19,
  },
  reviewBtn: {
    padding: variables.contentPadding,
  },
})

export default WithdrawCeloReviewScreen
