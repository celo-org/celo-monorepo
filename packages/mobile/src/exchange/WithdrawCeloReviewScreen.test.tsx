import BigNumber from 'bignumber.js'
import * as React from 'react'
import 'react-native'
import { fireEvent, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import { Actions } from 'src/exchange/actions'
import WithdrawCeloReviewScreen from 'src/exchange/WithdrawCeloReviewScreen'
import { Screens } from 'src/navigator/Screens'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

const SAMPLE_ADDRESS = '0xcc642068bdbbdeb91f348213492d2a80ab1ed23c'
const SAMPLE_AMOUNT = new BigNumber(5.001)

const mockScreenProps = getMockStackScreenProps(Screens.WithdrawCeloReviewScreen, {
  recipientAddress: SAMPLE_ADDRESS,
  amount: SAMPLE_AMOUNT,
  feeEstimate: new BigNumber(1),
})

const store = createMockStore()

describe('WithdrawCeloReviewScreen', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <WithdrawCeloReviewScreen {...mockScreenProps} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('dispatches a withdraw event when the button is clicked', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <WithdrawCeloReviewScreen {...mockScreenProps} />
      </Provider>
    )

    fireEvent.press(getByTestId('ConfirmWithdrawButton'))
    expect(store.getActions()).toEqual([
      {
        type: Actions.WITHDRAW_CELO,
        amount: SAMPLE_AMOUNT,
        recipientAddress: SAMPLE_ADDRESS,
      },
    ])
  })
})
