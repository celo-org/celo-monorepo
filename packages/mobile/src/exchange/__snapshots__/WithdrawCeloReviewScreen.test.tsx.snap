// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithdrawCeloReviewScreen renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 38,
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 5,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginRight": 5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
            ]
          }
        >
          exchangeFlow9:withdrawCeloTo
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 19,
              "lineHeight": 22,
            }
          }
        >
          0xcc64...d23c
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "borderStyle": "solid",
          "borderTopColor": "#F8F9F9",
          "borderTopWidth": 1,
          "marginBottom": 15,
          "marginTop": 10,
          "width": "100%",
        }
      }
    />
    <View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 5,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
              ]
            }
          >
            exchangeFlow9:withdrawCeloAmount
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
          >
            
            
            5.001
          </Text>
        </Text>
      </View>
      <View>
        <View
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="feeDrawer/WithdrawCelo"
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                feeEstimate
              </Text>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <svg
                  fill="none"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    d="M3 6l5 5 5-5"
                    stroke="#B4B9BD"
                  />
                </svg>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "marginVertical": 5,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginRight": 5,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Inter-Regular",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      undefined,
                    ]
                  }
                >
                  
                </Text>
              </View>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 0.7,
                      },
                    ],
                  }
                }
              >
                <ActivityIndicator
                  animating={true}
                  color="#42D689"
                  hidesWhenStopped={true}
                  size="small"
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 5,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
              ]
            }
          >
            total
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
          >
            
            
            5.001
          </Text>
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 5,
            },
            Object {
              "marginVertical": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
              ]
            }
          >
            Celo Dollars @
             
            <Text
              numberOfLines={1}
              style={
                Array [
                  undefined,
                  Object {
                    "color": undefined,
                  },
                ]
              }
            >
              
              $
              0.75
            </Text>
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#9CA4A9",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
          >
            
            5.001
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
        },
        Object {
          "padding": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 100,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#EEB93C",
            "flexGrow": 1,
            "height": 48,
            "justifyContent": "center",
            "opacity": undefined,
            "paddingHorizontal": 24,
            "paddingVertical": 5,
          }
        }
        testID="ConfirmWithdrawButton"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          global:withdraw
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
